src/act.other.cpp

```diff
@@ -4165,6 +4165,128 @@ ACMD(do_memory)
   }
 }
 
+// Looks for any working cyberdeck in the character's inventory or worn equipment.
+struct obj_data *find_cyberdeck(struct char_data *ch)
+{
+  struct obj_data *deck;
+  if (!ch) {
+    mudlog("SYSERR: Got NULL ch to find_cyberdeck()!", ch, LOG_SYSLOG, TRUE);
+    return NULL;
+  }
+
+  for (deck = ch->carrying; deck; deck = deck->next_content) {
+    if (OBJ_IS_VALID_CYBERDECK(deck)) {
+      return deck;
+    }
+  }
+
+  for (int wear_idx = 0; !deck && wear_idx < NUM_WEARS; wear_idx++) {
+    if (OBJ_IS_VALID_CYBERDECK(GET_EQ(ch, wear_idx))) {
+      return GET_EQ(ch, wear_idx);
+    }
+  }
+
+  return NULL;
+}
+
+ACMD(do_download_headware)
+{
+  struct obj_data *deck, *jack, *memory;
+  skip_spaces(&argument);
+  if (!*argument) {
+    send_to_char("Which file number do you want to transfer out of headware memory?\r\n", ch);
+    return;
+  }
+
+  for (memory = ch->cyberware; memory; memory = memory->next_content) {
+    if (GET_OBJ_VAL(memory, 0) == CYB_MEMORY) {
+      break;
+    }
+  }
+  if (!memory) {
+    send_to_char("You have no headware memory.\r\n", ch);
+    return;
+  }
+
+  // Error message sent in function.
+  if (!(jack = get_datajack(ch, FALSE))) {
+    return;
+  }
+
+  if (!(deck = find_cyberdeck(ch))) {
+    send_to_char(ch, "You need to have a working cyberdeck to transfer data out of headware memory.\r\n");
+    return;
+  }
+
+  int x = atoi(argument);
+  struct obj_data *obj = memory->contains;
+  for (; obj; obj = obj->next_content) {
+    if (!--x)
```

Comment:
If a user gives you `download -1`, you'll have some pretty nasty behavior here. Recommend an error case after the `atoi()` call, something like `FAILURE_CASE(x <= 0, "You must provide a positive integer.")`

You may also need to switch to `x--` rather than `--x`, unless you're listing a zero-indexed list to the player with a +1 index. I think the current state is that `download 1` would get file 0, and `download 0` would run into the negatives.

Reply:
This snippet was pulled from the code in `do_delete`, it seems to work as intended but I agree it's not perfect, I'll try rewriting it

Reply:
I think adding the `FAILURE_CASE` is sufficient, the list is zero indexed but displayed starting with 1.

```c
  int x = atoi(argument);
  FAILURE_CASE(x <= 0, "You must provide a positive file index number.")
  struct obj_data *obj = memory->contains;
  for (; obj; obj = obj->next_content) {
    if (!--x)
      break;
  }
  ```